File: src/android/libs/SmartSync/src/com/salesforce/androidsdk/smartsync/util/SyncDownTarget.java
/*
 * Copyright (c) 2014, salesforce.com, inc.
 * All rights reserved.
 * Redistribution and use of this software in source and binary forms, with or
 * without modification, are permitted provided that the following conditions
 * are met:
 * - Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 * - Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 * - Neither the name of salesforce.com, inc. nor the names of its contributors
 * may be used to endorse or promote products derived from this software without
 * specific prior written permission of salesforce.com, inc.
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
package com.salesforce.androidsdk.smartsync.util;

import android.util.Log;

import com.salesforce.androidsdk.smartsync.manager.SyncManager;
import com.salesforce.androidsdk.util.JSONObjectHelper;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.lang.reflect.Constructor;

/**
 * Target for sync down:
 * - what records to download from server
 * - how to download those records
 */
public abstract class SyncDownTarget extends SyncTarget {

    // Constants
	public static final String QUERY_TYPE = "type";

    // Fields
	protected QueryType queryType;
    protected int totalSize; // set during a fetch

    /**
	 * Build SyncDownTarget from json
	 * @param target as json
	 * @return
	 * @throws JSONException
	 */
	@SuppressWarnings("unchecked")
	public static SyncDownTarget fromJSON(JSONObject target) throws JSONException {
		if (target == null)
			return null;

		QueryType queryType = QueryType.valueOf(target.getString(QUERY_TYPE));

        switch (queryType) {
        case mru:     return new MruSyncDownTarget(target);
        case sosl:    return new SoslSyncDownTarget(target);
        case soql:    return new SoqlSyncDownTarget(target);
        case custom:
        default:
            try {
                Class<? extends SyncDownTarget> implClass = (Class<? extends SyncDownTarget>) Class.forName(target.getString(ANDROID_IMPL));
                Constructor<? extends SyncDownTarget> constructor = implClass.getConstructor(JSONObject.class);
                return constructor.newInstance(target);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }
	}

    /**
     * Construct SyncDownTarget
     */
    public SyncDownTarget() {
        super();
    }

    /**
     * Construct SyncDownTarget from json
     * @param target
     * @throws JSONException
     */
    public SyncDownTarget(JSONObject target) throws JSONException {
        super(target);
        queryType = QueryType.valueOf(target.getString(QUERY_TYPE));
    }

    /**
     * @return json representation of target
     * @throws JSONException
     */
    public JSONObject asJSON() throws JSONException {
        JSONObject target = super.asJSON();
        target.put(QUERY_TYPE, queryType.name());
        return target;
    }

    /**
     * Start fetching records conforming to target
     * If a value for maxTimeStamp greater than 0 is passed in, only records created/modified after maxTimeStamp should be returned
     * @param syncManager
     * @param maxTimeStamp
     * @throws IOException, JSONException
     */
    public abstract JSONArray startFetch(SyncManager syncManager, long maxTimeStamp) throws IOException, JSONException;

    /**
     * Continue fetching records conforming to target if any
     * @param syncManager
     * @return null if there are no more records to fetch
     * @throws IOException, JSONException
     */
    public abstract JSONArray continueFetch(SyncManager syncManager) throws IOException, JSONException;

    /**
     * @return number of records expected to be fetched - is set when startFetch() is called
     */
    public int getTotalSize() {
        return totalSize;
    }

    /**
     * @return QueryType of this target
     */
    public QueryType getQueryType() {
        return queryType;
    }


    /**
     * Gets the latest modification timestamp from the array of records.
     * @param records
     * @return latest modification time stamp
     * @throws JSONException
     */
    public long getLatestModificationTimeStamp(JSONArray records) throws JSONException {
        long maxTimeStamp = -1;
        for (int i = 0; i < records.length(); i++) {
            String timeStampStr = JSONObjectHelper.optString(records.getJSONObject(i), getModificationDateFieldName());
            if (timeStampStr == null) {
                maxTimeStamp = -1;
                break; // field not present
            }
            try {
                long timeStamp = Constants.TIMESTAMP_FORMAT.parse(timeStampStr).getTime();
                maxTimeStamp = Math.max(timeStamp, maxTimeStamp);
            } catch (Exception e) {
                Log.w("SyncDownTarget.getLatestModificationTimeStamp", "Could not parse modification date field " + getModificationDateFieldName(), e);
                maxTimeStamp = -1;
                break;
            }
        }
        return maxTimeStamp;
    }

    /**
     * Enum for query type
     */
    public enum QueryType {
    	mru,
    	sosl,
    	soql,
        custom
    }
}

File: src/windows/src/SmartSync/Model/SyncDownTarget.cs
/*
 * Copyright (c) 2015, salesforce.com, inc.
 * All rights reserved.
 * Redistribution and use of this software in source and binary forms, with or
 * without modification, are permitted provided that the following conditions
 * are met:
 * - Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 * - Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 * - Neither the name of salesforce.com, inc. nor the names of its contributors
 * may be used to endorse or promote products derived from this software without
 * specific prior written permission of salesforce.com, inc.
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

using System;
using System.Diagnostics;
using System.Reflection;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using Newtonsoft.Json.Linq;
using Salesforce.SDK.SmartStore.Store;
using Salesforce.SDK.SmartSync.Manager;
using Salesforce.SDK.SmartSync.Util;

namespace Salesforce.SDK.SmartSync.Model
{
    public abstract class SyncDownTarget : SyncTarget
    {
        public enum QueryTypes
        {
            Mru,
            Sosl,
            Soql,
            Custom
        }
        public new const string WindowsImpl = "windowsImpl";
        public new const string WindowsImplType = "windowsImplType";
        public const string Type = "type";
        public const string QueryString = "query";
        public new const string ModificationDateFieldName = "modificationDateFieldName";
        public const int Unchanged = -1;

        public QueryTypes QueryType { protected set; get; }

        public string Query { protected set; get; }
        
        public int TotalSize { protected set; get; } // set during fetch

        /// <summary>
        ///     Build SyncTarget from json
        /// </summary>
        /// <param name="target"></param>
        /// <returns></returns>
        public static SyncDownTarget FromJson(JObject target)
        {
            if (target == null) return null;
            JToken impl;
            if (target.TryGetValue(WindowsImpl, out impl))
            {
                JToken implType;
                if (target.TryGetValue(WindowsImplType, out implType))
                {
                    try
                    {
                        Assembly assembly = Assembly.Load(new AssemblyName(impl.ToObject<string>()));
                        Type type = assembly.GetType(implType.ToObject<string>());
                        if (type.GetTypeInfo().IsSubclassOf(typeof(SyncTarget)))
                        {
                            MethodInfo method = type.GetTypeInfo().GetDeclaredMethod("FromJson");
                            return (SyncDownTarget)method.Invoke(type, new object[] { target });
                        }
                    }
                    catch (Exception)
                    {
                        throw new SmartStoreException("Invalid SyncDownTarget");
                    }
                }
            }
            throw new SmartStoreException("Could not generate SyncDownTarget from json target");
        }

        public static string AddFilterForReSync(string query, long maxTimeStamp)
        {
            if (maxTimeStamp != Unchanged)
            {
                string extraPredicate = Constants.LastModifiedDate + " > " +
                                        new DateTime(maxTimeStamp, DateTimeKind.Utc).ToString("o");
                if (query.Contains(" where "))
                {
                    var reg = new Regex("( where )");
                    query = reg.Replace(query, "$1 where " + extraPredicate + " and ", 1);
                }
                else
                {
                    string pred = "$1 where " + extraPredicate;
                    var reg = new Regex("( from[ ]+[^ ]*)");
                    query = reg.Replace(query, pred, 1);
                }
            }
            return query;
        }

        public override JObject AsJson()
        {
            var target = base.AsJson();
            target[Type] = QueryType.ToString();
            target[QueryString] = Query;
            return target;
        }

        protected SyncDownTarget(JObject target)
            : base(target)
        {
            QueryType = (QueryTypes)Enum.Parse(typeof(QueryTypes), target.ExtractValue<string>(Constants.QueryType));
        }

        protected SyncDownTarget()
        {

        }

        protected SyncDownTarget(string query)
        {
            Query = query;
        }

        public abstract Task<JArray> StartFetch(SyncManager syncManager, long maxTimeStamp);

        public abstract Task<JArray> ContinueFetch(SyncManager syncManager);

        /// <summary>
        /// Method to return number of records to be fetched
        /// </summary>
        /// <returns>Number of records expected to be fetched - is set when startFetch() is called</returns>
        public int GetTotalSize()
        {
            return TotalSize;
        }


        public long GetMaxTimeStamp(JArray jArray)
        {
            long maxTimeStamp = Unchanged;
            foreach (JToken t in jArray)
            {
                var jObj = t.ToObject<JObject>();
                if (jObj != null)
                {
                    var date = jObj.ExtractValue<string>(GetModificationDate());
                    if (String.IsNullOrWhiteSpace(date))
                    {
                        maxTimeStamp = Unchanged;
                        break;
                    }
                    try
                    {
                        long timeStamp = Convert.ToDateTime(date).Ticks;
                        maxTimeStamp = Math.Max(timeStamp, maxTimeStamp);
                    }
                    catch (Exception)
                    {
                        Debug.WriteLine("SmartSync.GetMaxTimeStamp could not parse LastModifiedDate");
                        maxTimeStamp = Unchanged;
                        break;
                    }
                }
            }
            return maxTimeStamp;
        }
    }
}
